cluster-autoscaler/simulator/predicate_error.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	"fmt"
+	"strings"
+)
+
+// PredicateErrorType is type of predicate error
+type PredicateErrorType int
+
+const (
+	// NotSchedulablePredicateError means that one of the filters retuned that pod does not fit a node
+	NotSchedulablePredicateError PredicateErrorType = iota
+	// InternalPredicateError denotes internal unexpected error while calling PredicateChecker
+	InternalPredicateError
+)
+
+// PredicateError represents an error during predicate checking.
+type PredicateError *predicateError
+
+type predicateError struct {
+	errorType     PredicateErrorType
+	predicateName string
+	errorMessage  string
+	reasons       []string
+	// debugInfo contains additional info that predicate doesn't include,
+	// but may be useful for debugging (e.g. taints on node blocking scale-up)
+	debugInfo func() string
+}
+
+// PredicateName return name of predicate which failed.
+func (pe *predicateError) ErrorType() PredicateErrorType {
+	return pe.errorType
+}
+
+// PredicateName return name of predicate which failed.
+func (pe *predicateError) PredicateName() string {
+	return pe.predicateName
+}
+
+// Message returns error message.
+func (pe *predicateError) Message() string {
+	if pe.errorMessage == "" {
+		return "unknown error"
```

Comment:
nit: I sort of dislike this error message. It implies something bad has happened, but in reality errorMessage == "" usually means we don't care for the actual error message (my reaction to seeing the previous 'Predicates failed' in logs would be less nervous than if it was 'unknown error').

Reply:
Don't buy it. Generally we should not construct `PredicateError` with empty message. If we have scary default message we will incentify fixing the code which poorly create `PredicateError`

Reply:
I thought we do this all the time to avoid sprintf costs on throwaway errors? Or do we set some generic string when we generate those?